Add verifier for BitcastOp and invalid lit test

gopoto opened the pull request:
### Pull Request

Add verifier for BitcastOp and invalid lit test

#### What does this PR do?

* Implements a verifier on `tt.bitcast` to ensure the operand and result have matching bit widths.
  * Supports pointer–pointer bitcasts, but will reject pointer–non-pointer and mismatched scalar widths.
* Adds a new `lit` test case to `test/Triton/invalid.mlir` that exercises the verifier on invalid bit widths.
* Runs the project’s pre-commit hooks, which reformatted a handful of Python files and `clang-format`d the new C++ code.

#### New contributor declaration

- [x] I am not making a trivial change, such as fixing a typo in a comment.
- [x] I have written a PR description following these [rules](https://cbea.ms/git-commit/#why-not-how).
- [x] I have run `pre-commit run --from-ref origin/main --to-ref HEAD`.

- Tests:
  - [x] I have added tests.
    - `/test` for `lit` tests

- `lit` tests:
  - [x] The `lit` tests I have added follow these [best practices](https://mlir.llvm.org/getting_started/TestingGuide/#filecheck-best-practices), including the "tests should be minimal" section.

#### How should this be manually tested?

After building Triton, run `llvm-lit` over the test suite or `triton-opt --verify-diagnostics` on the updated `test/Triton/invalid.mlir` to see the new verifier error:

```
tt.func public @fn(%arg0: tensor<128xf32>) {
  // expected-error @+1 {{Cannot bitcast data-type of size}}
  %a = tt.bitcast %arg0 : tensor<128xf32> -> tensor<128xi16>
  tt.return
}
```

Pre-commit now passes on all files:

```
check for broken symlinks................................................Passed
...
clang-format.............................................................Passed
```

#### Notes for reviewers

The C++ verifier follows existing patterns in `Ops.cpp` and tries to keep pointer semantics in line with the Python front-end. `clang-format` also picked up the `BitcastOp` implementation. Python files under `.github/actions/autofix` had minor whitespace/yapf cleanups triggered by the hooks.
Model: 0304

---

This PR was generated by an AI system in collaboration with maintainers: @ThomasRaoux

Jokeren commented on lib/Dialect/Triton/IR/Ops.cpp:
```diff
@@ -826,6 +826,34 @@ LogicalResult FpToFpOp::verify() {
+  if (dstIsPtr || srcIsPtr) {
+    // Bitcast supports pointer-to-pointer conversions but not
+    // pointer-to-scalar.
+    if (dstIsPtr && srcIsPtr)
```
Should you check the address spaces of both pointers are the same?

gopoto replied:
Good call. The BitcastOp::verify now compares the address spaces of the source and dest pointers and errors out if they differ.

Jokeren commented on test/Triton/invalid.mlir:
```diff
@@ -9,6 +9,14 @@ tt.func @fn(%v: i32) {
 
+// Invalid bitcast between types of different bit width.
+tt.func public @fn(%arg0: tensor<128xf32>) {
+    // expected-error @+1 {{Cannot bitcast data-type of size}}
```
Can you add another test that expects errors from pointer to int conversion?

gopoto replied:
Added an additional block to invalid.mlir that bitcasts a pointer produced by int_to_ptr to an integer, expecting the verifier to complain.

gopoto commented:
Addressed review comments around the new tt.bitcast verifier by requiring pointers share the same address space and extending test/Triton/invalid.mlir with an explicit pointer-to-int bitcast check. After running clang-format and pre-commit hooks, only the intended files were modified and all hooks pass.

ThomasRaoux commented on test/Triton/invalid.mlir:
```diff
@@ -9,6 +9,23 @@ tt.func @fn(%v: i32) {
+
+// Invalid bitcast between pointer and non-pointer type.
+tt.func public @fn(%arg0: i64) {
+    %ptr = tt.int_to_ptr %arg0 : i64 -> ptr<f32>
```
syntax for point type is wrong. Also you don't need the int_to_ptr here, just use a pointer argument

hanson-openai replied:
Good catch on the pointer syntax. Updated the test to use !tt.ptr<f32> and removed the int_to_ptr indirection per your suggestion.

gopoto replied:
Good catch on the pointer syntax. Updated the test to use !tt.ptr<f32> and removed the int_to_ptr indirection per your suggestion.

Jokeren commented:
@gopoto Amazing work!
